superset/utils/decorators.py

```diff
@@ -29,3 +35,60 @@ def stats_timing(stats_key, stats_logger):
         raise e
     finally:
         stats_logger.timing(stats_key, now_as_float() - start_ts)
+
+
+def etag_cache(max_age, check_perms=bool):
+    """
+    A decorator for caching views and handling etag conditional requests.
+
+    The decorator caches the response, returning headers for etag and last
+    modified. If the client makes a request that matches, the server will
+    return a "304 Not Mofified" status.
+
+    If no cache is set, the decorator will still set the ETag header, and
+    handle conditional requests.
+
+    """
+    def decorator(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            # check if the user can access the resource
+            check_perms(*args, **kwargs)
+
+            try:
+                # build the cache key from the function arguments and any other
+                # additional GET arguments (like `form_data`, eg).
+                key_args = list(args)
+                key_kwargs = kwargs.copy()
+                key_kwargs.update(request.args)
+                cache_key = wrapper.make_cache_key(f, *key_args, **key_kwargs)
+                response = cache.get(cache_key)
+            except Exception:  # pylint: disable=broad-except
+                if app.debug:
+                    raise
+                logging.exception('Exception possibly due to cache backend.')
+                response = None
+
+            if response is None or request.method == 'POST':
+                response = f(*args, **kwargs)
+                response.cache_control.public = True
+                response.last_modified = datetime.utcnow()
+                response.expires = response.last_modified + timedelta(seconds=max_age)
+                response.add_etag()
+                try:
+                    cache.set(cache_key, response, timeout=max_age)
```

Comment:
@betodealmeida note in [Flask-Caching](https://github.com/sh4nks/flask-caching/blob/5599d7aafdfa36f1f3cfabfd19a13942c98352d4/flask_caching/backends/simple.py#L19), 

> A timeout of 0 indicates that the cache never expires.

which is inconsistent with the result expiry if the `max_age` is zero, i.e., the backend cache never expires whereas the response cache immediately expires.

Reply:
@john-bodley good point. I'll fix this.